kv/kvserver: TestStrictGCEnforcement failed

[(kv/kvserver).TestStrictGCEnforcement failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2697767&tab=buildLog) on [master@8b6f3c84cc256debeeb4d4055c5f0d5c9a481213](https://github.com/cockroachdb/cockroach/commits/8b6f3c84cc256debeeb4d4055c5f0d5c9a481213):

```
=== RUN   TestStrictGCEnforcement
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStrictGCEnforcement865385043
    test_log_scope.go:74: use -show-logs to present logs inline
=== CONT  TestStrictGCEnforcement
    client_replica_test.go:3277: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStrictGCEnforcement865385043
--- FAIL: TestStrictGCEnforcement (0.58s)
=== RUN   TestStrictGCEnforcement/protected_timestamps_are_respected
    client_replica_test.go:3275: 
        	Error Trace:	client_replica_test.go:3148
        	            				client_replica_test.go:3275
        	Error:      	Received unexpected error:
        	            	batch timestamp 1613803810.475835111,0 must be after replica GC threshold 1613803819.542992787,0
        	Test:       	TestStrictGCEnforcement/protected_timestamps_are_respected
E210220 06:50:20.549634 490766 kv/kvclient/kvcoord/txn_interceptor_committer.go:449  [-] 1  making txn commit explicit failed for "sql txn" meta={id=e4613806 pri=0.00440269 epo=0 ts=1613803820.543560449,0 min=1613803820.543560449,0 seq=6} lock=true stat=STAGING rts=1613803820.543560449,0 wto=false gul=1613803821.043560449,0 ifw=5: result is ambiguous (server shutdown)
    --- FAIL: TestStrictGCEnforcement/protected_timestamps_are_respected (0.02s)

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-parallel=4

```
make stressrace TESTS=TestStrictGCEnforcement PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestStrictGCEnforcement.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(kv/kvserver).TestStrictGCEnforcement failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2698559&tab=buildLog) on [master@8b6f3c84cc256debeeb4d4055c5f0d5c9a481213](https://github.com/cockroachdb/cockroach/commits/8b6f3c84cc256debeeb4d4055c5f0d5c9a481213):

```
=== RUN   TestStrictGCEnforcement
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStrictGCEnforcement325833239
    test_log_scope.go:74: use -show-logs to present logs inline
=== CONT  TestStrictGCEnforcement
    client_replica_test.go:3277: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStrictGCEnforcement325833239
--- FAIL: TestStrictGCEnforcement (1.20s)
=== RUN   TestStrictGCEnforcement/protected_timestamps_are_respected
    client_replica_test.go:3275: 
        	Error Trace:	client_replica_test.go:3148
        	            				client_replica_test.go:3275
        	Error:      	Received unexpected error:
        	            	batch timestamp 1613805982.860039020,0 must be after replica GC threshold 1613805992.034763111,0
        	Test:       	TestStrictGCEnforcement/protected_timestamps_are_respected
    --- FAIL: TestStrictGCEnforcement/protected_timestamps_are_respected (0.06s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestStrictGCEnforcement PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestStrictGCEnforcement.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(kv/kvserver).TestStrictGCEnforcement failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2700615&tab=buildLog) on [master@d485973754bd65c2b504b47d0c22a512b6ce00e0](https://github.com/cockroachdb/cockroach/commits/d485973754bd65c2b504b47d0c22a512b6ce00e0):

```
=== RUN   TestStrictGCEnforcement
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStrictGCEnforcement525480173
    test_log_scope.go:74: use -show-logs to present logs inline
=== CONT  TestStrictGCEnforcement
    client_replica_test.go:3277: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStrictGCEnforcement525480173
--- FAIL: TestStrictGCEnforcement (1.96s)
=== RUN   TestStrictGCEnforcement/protected_timestamps_are_respected
    client_replica_test.go:3275: 
        	Error Trace:	client_replica_test.go:3148
        	            				client_replica_test.go:3275
        	Error:      	Received unexpected error:
        	            	batch timestamp 1613909497.453314447,0 must be after replica GC threshold 1613909506.607677054,0
        	Test:       	TestStrictGCEnforcement/protected_timestamps_are_respected
E210221 12:11:47.636673 587356 kv/kvclient/kvcoord/txn_interceptor_committer.go:449  [-] 1  making txn commit explicit failed for "sql txn" meta={id=66368a96 pri=0.00539576 epo=0 ts=1613909507.626957051,0 min=1613909507.626957051,0 seq=6} lock=true stat=STAGING rts=1613909507.626957051,0 wto=false gul=1613909508.126957051,0 ifw=5: result is ambiguous (server shutdown)
    --- FAIL: TestStrictGCEnforcement/protected_timestamps_are_respected (0.05s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestStrictGCEnforcement PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestStrictGCEnforcement.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(kv/kvserver).TestStrictGCEnforcement failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2701253&tab=buildLog) on [master@682582fd65a512d90f187a3f7c8e368a43bd89e9](https://github.com/cockroachdb/cockroach/commits/682582fd65a512d90f187a3f7c8e368a43bd89e9):

```
=== RUN   TestStrictGCEnforcement
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStrictGCEnforcement976943093
    test_log_scope.go:74: use -show-logs to present logs inline
=== CONT  TestStrictGCEnforcement
    client_replica_test.go:3277: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStrictGCEnforcement976943093
--- FAIL: TestStrictGCEnforcement (1.00s)
=== RUN   TestStrictGCEnforcement/protected_timestamps_are_respected
    client_replica_test.go:3275: 
        	Error Trace:	client_replica_test.go:3148
        	            				client_replica_test.go:3275
        	Error:      	Received unexpected error:
        	            	batch timestamp 1613945891.628233339,0 must be after replica GC threshold 1613945900.727739693,0
        	Test:       	TestStrictGCEnforcement/protected_timestamps_are_respected
E210221 22:18:21.735737 587063 kv/kvclient/kvcoord/txn_interceptor_committer.go:449  [-] 1  making txn commit explicit failed for "sql txn" meta={id=0b8fb5b4 pri=0.00782274 epo=0 ts=1613945901.729156224,0 min=1613945901.729156224,0 seq=6} lock=true stat=STAGING rts=1613945901.729156224,0 wto=false gul=1613945902.229156224,0 ifw=5: result is ambiguous (server shutdown)
    --- FAIL: TestStrictGCEnforcement/protected_timestamps_are_respected (0.03s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestStrictGCEnforcement PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestStrictGCEnforcement.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(kv/kvserver).TestStrictGCEnforcement failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2701604&tab=buildLog) on [master@3825fd1139ee2ab25cf48da37d315e00156e61c1](https://github.com/cockroachdb/cockroach/commits/3825fd1139ee2ab25cf48da37d315e00156e61c1):

```
=== RUN   TestStrictGCEnforcement
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStrictGCEnforcement985101797
    test_log_scope.go:74: use -show-logs to present logs inline
=== CONT  TestStrictGCEnforcement
    client_replica_test.go:3277: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStrictGCEnforcement985101797
--- FAIL: TestStrictGCEnforcement (1.61s)
=== RUN   TestStrictGCEnforcement/protected_timestamps_are_respected
    client_replica_test.go:3275: 
        	Error Trace:	client_replica_test.go:3148
        	            				client_replica_test.go:3275
        	Error:      	Received unexpected error:
        	            	batch timestamp 1613953872.049883014,0 must be after replica GC threshold 1613953881.259591168,0
        	Test:       	TestStrictGCEnforcement/protected_timestamps_are_respected
    --- FAIL: TestStrictGCEnforcement/protected_timestamps_are_respected (0.05s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestStrictGCEnforcement PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestStrictGCEnforcement.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(kv/kvserver).TestStrictGCEnforcement failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2701696&tab=buildLog) on [master@894efb0215580cddc285e21b1232326b375d391c](https://github.com/cockroachdb/cockroach/commits/894efb0215580cddc285e21b1232326b375d391c):

```
=== RUN   TestStrictGCEnforcement
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStrictGCEnforcement343167773
    test_log_scope.go:74: use -show-logs to present logs inline
=== CONT  TestStrictGCEnforcement
    client_replica_test.go:3277: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStrictGCEnforcement343167773
--- FAIL: TestStrictGCEnforcement (1.25s)
=== RUN   TestStrictGCEnforcement/protected_timestamps_are_respected
    client_replica_test.go:3275: 
        	Error Trace:	client_replica_test.go:3148
        	            				client_replica_test.go:3275
        	Error:      	Received unexpected error:
        	            	batch timestamp 1613955961.312463123,0 must be after replica GC threshold 1613955970.413014901,0
        	Test:       	TestStrictGCEnforcement/protected_timestamps_are_respected
E210222 01:06:11.421804 587747 kv/kvclient/kvcoord/txn_interceptor_committer.go:449  [-] 1  making txn commit explicit failed for "sql txn" meta={id=52457830 pri=0.01948071 epo=0 ts=1613955971.413778932,0 min=1613955971.413778932,0 seq=6} lock=true stat=STAGING rts=1613955971.413778932,0 wto=false gul=1613955971.913778932,0 ifw=5: result is ambiguous (server shutdown)
    --- FAIL: TestStrictGCEnforcement/protected_timestamps_are_respected (0.03s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestStrictGCEnforcement PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestStrictGCEnforcement.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(kv/kvserver).TestStrictGCEnforcement failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2702033&tab=buildLog) on [master@6a95063b141824c8d98f248c1af87b1bd55bf435](https://github.com/cockroachdb/cockroach/commits/6a95063b141824c8d98f248c1af87b1bd55bf435):

```
=== RUN   TestStrictGCEnforcement
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStrictGCEnforcement078654418
    test_log_scope.go:74: use -show-logs to present logs inline
=== CONT  TestStrictGCEnforcement
    client_replica_test.go:3277: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStrictGCEnforcement078654418
--- FAIL: TestStrictGCEnforcement (1.08s)
=== RUN   TestStrictGCEnforcement/protected_timestamps_are_respected
    client_replica_test.go:3275: 
        	Error Trace:	client_replica_test.go:3148
        	            				client_replica_test.go:3275
        	Error:      	Received unexpected error:
        	            	batch timestamp 1613967361.583271845,0 must be after replica GC threshold 1613967370.677121380,0
        	Test:       	TestStrictGCEnforcement/protected_timestamps_are_respected
E210222 04:16:11.687066 582672 kv/kvclient/kvcoord/txn_interceptor_committer.go:449  [-] 1  making txn commit explicit failed for "sql txn" meta={id=b90ab80b pri=0.00754553 epo=0 ts=1613967371.678167503,0 min=1613967371.678167503,0 seq=6} lock=true stat=STAGING rts=1613967371.678167503,0 wto=false gul=1613967372.178167503,0 ifw=5: failed to send RPC: sending to all replicas failed; last error: node unavailable; try another peer
    --- FAIL: TestStrictGCEnforcement/protected_timestamps_are_respected (0.03s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestStrictGCEnforcement PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestStrictGCEnforcement.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(kv/kvserver).TestStrictGCEnforcement failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2702601&tab=buildLog) on [master@bf9744bad5a416a4b06907f0f3dd42896f7342f3](https://github.com/cockroachdb/cockroach/commits/bf9744bad5a416a4b06907f0f3dd42896f7342f3):

```
=== RUN   TestStrictGCEnforcement
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStrictGCEnforcement674214809
    test_log_scope.go:74: use -show-logs to present logs inline
=== CONT  TestStrictGCEnforcement
    client_replica_test.go:3277: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStrictGCEnforcement674214809
--- FAIL: TestStrictGCEnforcement (0.67s)
=== RUN   TestStrictGCEnforcement/protected_timestamps_are_respected
    client_replica_test.go:3275: 
        	Error Trace:	client_replica_test.go:3148
        	            				client_replica_test.go:3275
        	Error:      	Received unexpected error:
        	            	batch timestamp 1613976586.311069065,0 must be after replica GC threshold 1613976595.415678225,0
        	Test:       	TestStrictGCEnforcement/protected_timestamps_are_respected
E210222 06:49:56.431119 491414 kv/kvclient/kvcoord/txn_interceptor_committer.go:449  [-] 1  making txn commit explicit failed for "sql txn" meta={id=b5c110b9 pri=0.00208076 epo=0 ts=1613976596.416575923,0 min=1613976596.416575923,0 seq=6} lock=true stat=STAGING rts=1613976596.416575923,0 wto=false gul=1613976596.916575923,0 ifw=5: result is ambiguous (server shutdown)
    --- FAIL: TestStrictGCEnforcement/protected_timestamps_are_respected (0.03s)

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-parallel=4

```
make stressrace TESTS=TestStrictGCEnforcement PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestStrictGCEnforcement.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/kv/kvserver/replica.go
+++ b/pkg/kv/kvserver/replica.go
@@ -1251,7 +1251,8 @@ func (r *Replica) checkExecutionCanProceed(
 			if !r.canServeFollowerReadRLocked(ctx, ba, err) {
 				return st, err
 			}
-			err = nil // ignoring error
+			err = nil                     // ignore error
+			st = kvserverpb.LeaseStatus{} // already empty for follower reads, but be explicit
 		}
 	}
 

--- a/pkg/kv/kvserver/replica_range_lease.go
+++ b/pkg/kv/kvserver/replica_range_lease.go
@@ -942,9 +942,9 @@ func (r *Replica) checkRequestTimeRLocked(now hlc.ClockTimestamp, reqTS hlc.Time
 
 // leaseGoodToGoRLocked verifies that the replica has a lease that is
 // valid, owned by the current replica, and usable to serve requests at
-// the specified timestamp. The method will return the lease status,
-// along with an error indicating whether any of these conditions is
-// unsatisfied. The lease status is either empty or fully populated.
+// the specified timestamp. The method will return the lease status if
+// these conditions are satisfied or an error if they are unsatisfied.
+// The lease status is either empty or fully populated.
 //
 // Latches must be acquired on the range before calling this method.
 // This ensures that callers are properly sequenced with TransferLease
@@ -978,7 +978,7 @@ func (r *Replica) leaseGoodToGoRLocked(
 	st := r.leaseStatusForRequestRLocked(ctx, now, reqTS)
 	if !st.IsValid() {
 		// Case (2): invalid lease.
-		return st, false, &roachpb.InvalidLeaseError{}
+		return kvserverpb.LeaseStatus{}, false, &roachpb.InvalidLeaseError{}
 	}
 	if !st.Lease.OwnedBy(r.store.StoreID()) {
 		// Case (3): not leaseholder.
@@ -1022,7 +1022,7 @@ func (r *Replica) leaseGoodToGoRLocked(
 		}
 		// Otherwise, if the lease is currently held by another replica, redirect
 		// to the holder.
-		return st, false, newNotLeaseHolderError(
+		return kvserverpb.LeaseStatus{}, false, newNotLeaseHolderError(
 			st.Lease, r.store.StoreID(), r.descRLocked(), "lease held by different store",
 		)
 	}

--- a/pkg/kv/kvserver/replica_send.go
+++ b/pkg/kv/kvserver/replica_send.go
@@ -702,7 +702,7 @@ func (r *Replica) collectSpans(
 type endCmds struct {
 	repl *Replica
 	g    *concurrency.Guard
-	st   kvserverpb.LeaseStatus
+	st   kvserverpb.LeaseStatus // empty for follower reads
 }
 
 // move moves the endCmds into the return value, clearing and making a call to

